kmod/patch/kpatch-macros.h

```diff
@@ -39,6 +27,84 @@ struct kpatch_unload {
 #define KPATCH_IGNORE_FUNCTION(_fn) \
 	void *__kpatch_ignore_func_##_fn __section(.kpatch.ignore.functions) = _fn;
 
+
+/* Support for livepatch callback hooks */
+#if IS_ENABLED(CONFIG_LIVEPATCH)
+# if LINUX_VERSION_CODE >= KERNEL_VERSION(4, 15, 0)
+#  define HAS_LIVEPATCH_HOOKS
+# endif
+#endif
+
+#ifdef HAS_LIVEPATCH_HOOKS
+
+#include <linux/livepatch.h>
+
+typedef int (*kpatch_pre_patchcall_t)(struct klp_object *obj);
+typedef void (*kpatch_post_patchcall_t)(struct klp_object *obj);
+typedef void (*kpatch_pre_unpatchcall_t)(struct klp_object *obj);
+typedef void (*kpatch_post_unpatchcall_t)(struct klp_object *obj);
+
+struct kpatch_pre_patch {
+	kpatch_pre_patchcall_t fn;
+	char *objname; /* filled in by create-diff-object */
+};
+
+struct kpatch_post_patch {
+	kpatch_post_patchcall_t fn;
+	char *objname; /* filled in by create-diff-object */
+};
+
+struct kpatch_pre_unpatch {
+	kpatch_pre_unpatchcall_t fn;
+	char *objname; /* filled in by create-diff-object */
+};
+
+struct kpatch_post_unpatch {
+	kpatch_post_unpatchcall_t fn;
+	char *objname; /* filled in by create-diff-object */
+};
+
+
+#define KPATCH_PRE_PATCH_HOOK(_fn) \
```

Comment:
What do you think about converging our existing hook code to be more livepatch-esque?  For example: 

- Get rid of KPATCH_LOAD_HOOK for kpatch.ko in favor of the same KPATCH_{PRE,POST}_PATCH_HOOK interfaces we have for livepatch (and a similar change for the unload hook)

- Rename "hook" to "callback" throughout the code base

I think that would make things less confusing, to have the same terminology used by upstream livepatch, and also to have the same interfaces for kpatch and livepatch.

Reply:
Using the same terminology would help.  To be clear about the interface, you are suggesting that kpatch deprecate its hooks in lieu of adopting pre-and-post, patch-and-unpatch callbacks like livepatch?  I think this would definitely simplify the code (and documentation).

Reply:
> To be clear about the interface, you are suggesting that kpatch deprecate its hooks in lieu of adopting pre-and-post, patch-and-unpatch callbacks like livepatch?

Right.

Reply:
Still working on this (at glacial pace, sorry).  I've got the API sync'd, however there will be one interesting difference: for traditional kpatches, I don't think we can use the pre-patch status return value since they are invoked by way of module-notifier.  We can still check for non-zero status in the kpatch core, but maybe all we can do is issue a WARN on non-zero status?  Should we do something similar even in the kpatch pre-patch / vmlinux case?

Reply:
> Still working on this (at glacial pace, sorry). I've got the API sync'd, however there will be one interesting difference: for traditional kpatches, I don't think we can use the pre-patch status return value since they are invoked by way of module-notifier. We can still check for non-zero status in the kpatch core, but maybe all we can do is issue a WARN on non-zero status?

Yeah, I think it makes sense to WARN in the module notifier.

> Should we do something similar even in the kpatch pre-patch / vmlinux case?

I don't think so, might as well reject the patch there.